### ISSUE TITLE
Row Detail View plugin enhancement

### ISSUE THREAD

**Issue by ghiscoding** [ghiscoding/SlickGrid#1]
Let's talk about the Row Detail View in this issue

**Comment by SatanEnglish** [comment]
@ghiscoding
Ok ill try add the gridUID to the class name that way it's unique if multiple grids.

It doesn't necessarily have to be a large view to have that issue the gif's from my orginal view were not big.
I'll try get an example of tabs and resizing on tab changes if I have time.

**Comment by ghiscoding** [comment]
I made the changes described in previous post and cleanup a bit. The `querySelector` with `gridUID()` works great, so I don't think `[0]` is necessary anymore and I removed it. 

However, I now see what you describe in [comment](https://github.com/6pac/SlickGrid/pull/281#issuecomment-435274383) but even if I change to the code you suggested 
`_grid.getOptions().minRowBuffer = _options.panelRows + 3;`
That will not fix the issue, because with a super long View and I understand why, it's because SlickGrid doesn't hold enough rows in it's cache for the Row Detail to be recreated in time in the DOM. For example, if SlickGrid only has 15 rows (+ 3 or minRowBuffer, which makes it 15+3+3=21 rows) and your total amount of rows is say 30, then SlickGrid won't show anything until it passes 9 rows from the bottom. I might have completely misunderstood how it works, but I don't think I'm that far off with the logic.

**Comment by SatanEnglish** [comment]
document.querySelector("." + grid.getUID() + " .cellDetailView_" + item.id);
Nice that was exactly the thing I was thinking of but wasn't sure how to do I'll check it works with my other version when I'm at work on monday.

In regards to the `cellDetailView_x` issue I do not have a view bigger than the view port of the grid but have the same issue appearing. Think my setup is about 2/3 of the grid view port.

I'll try and describe my setup so that you can understand the hole picture.

What I have in my expanding view is a tab view. 
This Tab View has 4 tabs each of these tabs has more or less information inside it. 
Therefore for a nice user experience I am resizing the view on tab change to remove ugly masses of white space. Or optimize the view to take up as much space as possible.
This allows for most of my tabs to not have inner scrolling.
I then somewhere I have a limit on the detailView max-height (Maybe css) this is so that it doesn't every get bigger than the grid port.
The other thing I currently do on tab change is save the state so if it's scrolled back into view it is in the same state.

Todays been to bizzy but still hoping to get an example of tabs into the example.

**Comment by ghiscoding** [comment]
> Therefore for a nice user experience I am resizing the view on tab change to remove ugly masses of white space.

I understand what you're trying to do there, I think it's probably the best workable approach. What you could do instead of the `setTimeout` that you have on the RowDetailFormatter (what you wrote in the other open issue). I guess you could use `onScroll` that could loop through the array of expandable rows, and if the row comes back available (by checking the DOM element `cellDetailView_x` exist). Then you can resize that row detail view. I suggest you to do that looping in your code though, I don't think this belong to the plugin itself since it seems to be quite customized to your needs

**Comment by SatanEnglish** [comment]
@ghiscoding 
I've had a bit more of a play and now that I have tabs in there I can see that.
            // save the view when asked
            if (_options.saveDetailViewOnScroll) {
              saveDetailView(row);
            }

Inside the calculateOutOfRangeViews() function isn't working at all.

![2018-11-04_14-06-23](https://user-images.githubusercontent.com/2100563/47958938-de790c00-e03a-11e8-8d3b-27aa39357ede.gif)

With the handleOnScrollSaveDetailView(); un-commented
![2018-11-04_14-08-07](https://user-images.githubusercontent.com/2100563/47958947-27c95b80-e03b-11e8-9865-d9b8337c1c87.gif)

Any idea why the blocks not working how i'd expect.

Maybe I'll need to fire the saves off in your new calls BUT the save is still there in the code so I'd expect it to be working the same way

**Comment by SatanEnglish** [comment]
@ghiscoding 
PS have commited the tabs now so you can have a play.
I've got to go out with family for a while will try have another look myself when I have some time

**Comment by ghiscoding** [comment]
The `saveDetailViewOnScroll` is True by default and is a flag that I added so that I can disable the saveDetail because as I said earlier, I don't need or use this on my side since I use the framework (Angular/Aurelia) rendering engine. I set this flag to True so that it doesn't break your code. 

> Inside the calculateOutOfRangeViews() function isn't working at all.

Can you explain a bit more? I left your previous code of `handleScroll` (now renamed `handleOnScrollSaveDetailView`) in case my implementation of saving with `calculateOutOfRangeViews` doesn't work. So if the `handleOnScrollSaveDetailView` is still what works for you then put it back, that was Plan B. 

I would like to know, is the code worst or the same as before? My goal was to make it at least the same or better. I added the 2 subscribe methods that I need on my side, but again I use the framework (not the plugin, neither `saveDetailView`) for rendering the view, so I don't use it the same way as you do.

**Comment by ghiscoding** [comment]
I added a new option `outOfVisibleDetectEarlierRows` if you want to make an earlier detection. With this option set to `7` (in the example file), it seems to be working as your previous code `handleOnScrollSaveDetailView` does. Just pull the latest code to test it out.

**Comment by SatanEnglish** [comment]
@ghiscoding 
I'll give it a go today.
However if this does work might be worth just making it so that the `_options.saveDetailViewOnScroll` does something this under the hood. (As it doesn't do anything as is at the moment)

**Comment by SatanEnglish** [comment]
@ghiscoding

**Comment by SatanEnglish** [comment]
Ops didn't mean to close it.

**Comment by ghiscoding** [comment]
> PSSS: Found Y it's not firing `_gridRowBuffer` is never set anymore.

Ok I saw I deleted 2 lines by mistake in this [commit](https://github.com/ghiscoding/SlickGrid/commit/4af009ad391d3bb8485da26bfdbfd28d03d72a6b#diff-bd35064cc362e9c532cc856157980027L113) while I was supposed to replace both lines instead of deleting. Made a [commit](https://github.com/ghiscoding/SlickGrid/commit/634b4d0e900a26a66b71d6f268533c7292e49e37) that fixes it.

> PS: Concerned about your checkIsRowOutOfVisibleRange function shouldn't there be a check for the renderedRange.top seems your only check one direction of rows leaving the screen. (May just be not understanding what it's doing)

That's a good point, I previously always checked going down never up when implementing the code. So I made another [commit](https://github.com/ghiscoding/SlickGrid/commit/a2841afc9f74f0f843fbeefe76615b60d6d83a91) to fix that as well

**Comment by SatanEnglish** [comment]
Man you get stuff done quickly.

Ok I'll try have another play at my lunch break with the updates.

PS: Do you need some example code around how I handle filtering?

**Comment by ghiscoding** [comment]
Yes I'd like to get this plugin refactoring over with and shipped as I'd like to use it soon. After you test the new code, I think we will be able to safely remove the `handleOnScrollSaveDetailView()`. 

The filtering is part of the new example, I added the necessary padding handling. However I see that it probably needs a save of the view as well since it doesn't seem to keep the tab info. That is again something that I won't use on my side since I use the framework rendering engine. If you know how to fix it, then please make a commit.

**Comment by SatanEnglish** [comment]
Ok the scroll up and scroll down stuff isn't working.
I'm going to rework the calculations.

First off tho what is
` _visibleRenderedCellCount = renderedRange.bottom - renderedRange.top - _gridRowBuffer;`

Is this actually the visableRenderedROWCount??

**Comment by ghiscoding** [comment]
Seriously you should think about how you write your comments and questions. The way you write is a little frustrating, you're very often portraying that all of the code is broken. I passed a lot of time on this code rework (and you did too). It would be great if you could make the comments more constructive instead of always saying "all of it is broken... or all isn't working... or whatever else". There's a lot of the code that will be used by you and not on my side (for framework reasons that I've explained few times), however I still want to help in resolving all the issues (when possible). Perhaps it's just communication problems. Anyhow, it would be nice to make it fun for both of us. Thank you

To answer the questions
> _visibleRenderedCellCount = renderedRange.bottom - renderedRange.top - _gridRowBuffer;

That function is called `calculateVisibleRenderedCount()` and means what is rendered to the user (how many **rows** are visible on screen, so yes as you suggest we should rename to `_visibleRenderedRowCount` and add more `javadoc` to the function itself). 

However, by looking at the code, I realized that I made a small mistake `_gridRowBuffer` should be calculated twice, since there's a buffer on top **and** bottom. So the updated version should be
`_visibleRenderedCellCount = renderedRange.bottom - renderedRange.top - (_gridRowBuffer * 2);`

This doubling of `_gridRowBuffer` will most certainly affect the `checkIsRowOutOfVisibleRange()`, with the new numbers. If I set the example to `outOfVisibleDetectEarlierRows: 0`, it actually now call the `saveDetailView()` right in time. 

I won't do the commit until tomorrow, so if you wish to try out what I just wrote and possibly go ahead with your own commit, please go ahead. It's time for me to go to bed.

**Comment by SatanEnglish** [comment]
Sorry if my comments is frustrating. (I'll try and checking it doesn't sound like i'm saying it's all broken I do appreciate the amount of time you have put into this.)
I hate typing information back and forward as it always seems to getting miss understood.
I find one 5min chat on a call via something like discord or skype would save 2 days of going back and forward on messages.
This is really frustrating both end a few times now I've tried to give you exactly whats going wrong and it's getting understood wrong.

I tried
> _visibleRenderedCellCount = renderedRange.bottom - renderedRange.top - (_gridRowBuffer * 2);

Doesn't seem to be any better sorry.
I'm scrolling up and `notifyBackToVisibleWhenDomExist ` has been fired here is the values.
row = 5; 
rowIndex = 5;
renderedRange.top = 105;
_gridRowBuffer = 18;

This is quiet a bit out even scrolling at top speed there isn't any chance that row is in view range.

OK....  3 and a half hours later I may have found part of the issue!
Whenever a row is expand the renderedRange is increased as we have added rows....
This is likely messing with the calculation method.

Now while I do have some code that I am still work on in order to try and figure out a fix, right now its not working but I need to get home to bed.

Please Note the following code is not doing what it should yet.

Here is my over the top method.
It is still incomplete but may help for all the cases I have thought of so far in comments.
(Not including any fix for my find e.g the more rows issue)

        function calculateOutOfRangeViews() {
            if (_grid) {                
                var renderedRange = _grid.getRenderedRange();

                // Assume scroll direction is down by default.
                var scrollDir = 'DOWN';
                // If our new top is smaller we are scrolling up
                if (_lastRange && _lastRange.top > renderedRange.top) {
                    scrollDir = 'UP';
                }

                _expandedRows.forEach((row) => {
                    var rowIndex = _dataView.getRowById(row.id);

                    var isOutOfVisibility = checkIsRowOutOfVisibleRange(rowIndex, renderedRange, 0);

                    // save the view when asked
                    if (_options.saveDetailViewOnScroll) {
                        // If the top or bottom item of the range is an expanded row save it.
                        // This may need to be updated to check within current view port instead
                        if (rowIndex <= renderedRange.top + _gridRowBuffer || rowIndex >= renderedRange.bottom - _gridRowBuffer) {
                            saveDetailView(row);
                        }
                    }
                                        
                    if (scrollDir === 'UP') {
                        // If row index is above top but inside of the buffer area and we are scrolling up treat it like it's entering                    
                        if (rowIndex <= renderedRange.top && rowIndex > (renderedRange.top + _gridRowBuffer)) {
                            notifyBackToVisibleWhenDomExist(row, rowIndex);
                            console.log("BACK", rowIndex);
                        }
                                                
                        // Else if it's gone off the bottom since last grab of range
                        else if (rowIndex <= _lastRange.bottom && rowIndex > renderedRange.bottom) {
                            notifyOutOfVisibility(row, rowIndex);
                            console.log("NOTIFY", rowIndex);
                        }
                    }
                    else if (scrollDir === 'DOWN') {
                        var rowPadding = row[_keyPrefix + 'sizePadding'];

                        // If row index is below bottom but inside of the buffer area and we are scrolling down treat it like it's entering                    
                        if (rowIndex > renderedRange.bottom && rowIndex < (renderedRange.bottom - _gridRowBuffer)) {
                            notifyBackToVisibleWhenDomExist(row, rowIndex);
                            console.log("BACK", rowIndex);
                        }

                        // Else if it's gone off the top since last grab of range and it's expanded view part is also out of range
                        else if (rowPadding && rowIndex >= (_lastRange.top + rowPadding) && rowIndex < (renderedRange.top + rowPadding)) {
                            notifyOutOfVisibility(row, rowIndex);
                            console.log("NOTIFY", rowIndex);
                        }
                    }

                    //var isOutOfVisibility = checkIsRowOutOfVisibleRange(rowIndex, renderedRange, 0);

                    //if (!isOutOfVisibility && arrayFindIndex(_outOfVisibleRangeRows, rowIndex) >= 0) {
                    //    notifyBackToVisibleWhenDomExist(row, rowIndex);
                    //} else if (isOutOfVisibility) {
                    //    notifyOutOfVisibility(row, rowIndex);
                    //}
                });
                _lastRange = renderedRange;
            }
        }

I'll have another play tomorrow night as I cannot put anymore hours into it during work time.

**Comment by ghiscoding** [comment]
Thanks for understanding...

Wow this simple function is now becoming quite big. I won't have time to try it much until tonight. In your code, did you include the `(_gridRowBuffer * 2)` or not?

I was thinking also that the calculation I made most probably is only for the parent row which doesn't include the panel itself. I don't know if that makes a difference, but it's worth to know about. So if your panel is 15 rows long, the calculation will probably calculate the same way if you use a panel of 5 rows long.

**Comment by SatanEnglish** [comment]
My incomplete code is brute force so it's looking for  `_gridRowBuffer` on top and `_gridRowBuffer` on bottom. (Again it's not working yet as I ran out of time but it has all the test cases I use in it in comments)

I'd like to not have an extravagant function so if you can fix that simple calculation somehow that would be amazing. I'm not 100% sure what the clean up of rows does when we add so many by expanding rows tho.

As for the calculation of the panel being 15 the vs 5 rows long. As long as we treat the buffering range as if it's about going out of range it should be fine.

Also worth noting we can treat it going off the top different to it going off the bottom.
 Top can be as soon as parent goes out but before epxanded leaves range.
 e.g. bottom needs to be sooner as the buffering range will wipe expanded area pre-parent going out of range.
(All this should be handled by buffer space but it's still worth remembering)

**Comment by ghiscoding** [comment]
> As for the calculation of the panel being 15 the vs 5 rows long. As long as we treat the buffering range as if it's about going out of range it should be fine.

The calculation is more for knowing when the parent (the row where you clicked) goes out of the DOM. Basically if the parent element doesn't exist (which is the actual container of the panel), then your panel being 5 or 15 rows will be gone anyway since the parent got deleted. I just wanted to make sure that it's not part of your calculation in the other function.

P.S. I didn't try your code yet, I'm not even sure I'll have time today.

**Comment by SatanEnglish** [comment]
Think you may find going off the bottom is more forgiving than scrolling off the top here.

**Comment by SatanEnglish** [comment]
Ok here is a working version.

Again please by all means refactor it but the logic is correct now so you can at least see all the weird edge cases that have been accounted for.

I've likely still missed some cases that I haven't come up with but please open load of different views resize, don't resize, change tabs,

Scroll up, Scroll down, use some at the top of table scoll use some in middle, scroll use some at bottom.

Please not I have a hard coded 5 in for calls for checking that we are not trying to notify back in range on bottom items when we are scrolling in the top.

Again by all mean make this a variable this does **NOT** make sense to be the minBuffer before you ask.

I have posted the code here rather than commit it.
        
    var _lastRange = null;

        function calculateOutOfRangeViews() {
            if (_grid) {                
                var renderedRange = _grid.getRenderedRange();
               // Only check if we have expanded rows
                if (_expandedRows.length > 0) {
                    // Assume scroll direction is down by default.
                    var scrollDir = 'DOWN';
                    if (_lastRange) {
                        // If our new top is smaller we are scrolling up
                        if (_lastRange.top > renderedRange.top ||
                            // Or we are at very top but our bottom is increasing
                            (_lastRange.top === 0 && renderedRange.top === 0) && _lastRange.bottom > renderedRange.bottom) {
                            scrollDir = 'UP';
                        }

                        // Some scrolling isn't anything as the range is the same
                        if (_lastRange.top === renderedRange.top && _lastRange.bottom === renderedRange.bottom) {
                            return;
                        }
                    }
                }

                _expandedRows.forEach((row) => {
                    var rowIndex = _dataView.getRowById(row.id);

                    var rowPadding = row[_keyPrefix + 'sizePadding'];
                    var rowOutOfRange = arrayFindIndex(_outOfVisibleRangeRows, rowIndex) >= 0;

                    // save the view when asked
                    if (_options.saveDetailViewOnScroll) {
                        // If the top or bottom item of the range is an expanded row save it.
                        // This may need to be updated to check within current view port instead
                        if (rowIndex <= renderedRange.top + _gridRowBuffer || rowIndex >= renderedRange.bottom - _gridRowBuffer) {
                            saveDetailView(row);
                        }
                    }
                                        
                    if (scrollDir === 'UP') {
                        // If the row expanded area is within the buffer notify that it is back in range
                        if (rowOutOfRange && rowIndex - 5 < renderedRange.top && rowIndex >= renderedRange.top) {
                            notifyBackToVisibleWhenDomExist(row, rowIndex);
                            console.log("BACK", rowIndex);
                        }
                                                
                        // if our first expanded row is about to go off the bottom
                        if (!rowOutOfRange && (rowIndex + rowPadding) > renderedRange.bottom) {
                            notifyOutOfVisibility(row, rowIndex);
                            console.log("NOTIFY", rowIndex);
                        }
                    }
                    else if (scrollDir === 'DOWN') {
                        
                        // If row index is i higher than bottom with some added value (To ignore top rows off view) and is with view and was our of range                    
                        if (rowOutOfRange && (rowIndex + rowPadding + 5) > renderedRange.bottom && rowIndex < rowIndex + rowPadding) {
                            notifyBackToVisibleWhenDomExist(row, rowIndex);
                            console.log("BACK", rowIndex);
                        }

                        // if our row is outside top of and the buffering zone but not in the array of outOfVisable range notify it
                        if (!rowOutOfRange && rowIndex < renderedRange.top) {
                            notifyOutOfVisibility(row, rowIndex);
                            console.log("NOTIFY", rowIndex);
                        }
                    }

                    //var isOutOfVisibility = checkIsRowOutOfVisibleRange(rowIndex, renderedRange, 0);

                    //if (!isOutOfVisibility && arrayFindIndex(_outOfVisibleRangeRows, rowIndex) >= 0) {
                    //    notifyBackToVisibleWhenDomExist(row, rowIndex);
                    //} else if (isOutOfVisibility) {
                    //    notifyOutOfVisibility(row, rowIndex);
                    //}
                });
                _lastRange = renderedRange;
            }
        }

**Comment by ghiscoding** [comment]
Could you make a commit with your changes and I could test it later on my side? We can cleanup the process after once we find the working code. Thanks

Filter is the last thing to address right?

**Comment by SatanEnglish** [comment]
Filtering kind of works
> PS: Put the last of time before bed to get the filtering updated for you

I put in a case for saving on tab change I think. May need to put on explicitly for on filter call save on all expanded Rows that are in range??

**Comment by ghiscoding** [comment]
Tested the code on my side with my framework and it's still working fine. 

Some code review, there's a possible performance improvement that can be done on this [line](https://github.com/ghiscoding/SlickGrid/blob/bd83368fba49884d905d937180ccde4ce6fff9fb/plugins/slick.rowdetailview.js#L274) and this [line](https://github.com/ghiscoding/SlickGrid/blob/bd83368fba49884d905d937180ccde4ce6fff9fb/plugins/slick.rowdetailview.js#L288). I think both of these lines, we can change the `if` to an `else if`, because unless I misunderstood, if the first `if` is caught then it cannot go in the second `if`, right? 

Another possible improvement, it's always better to return early, there's no need to do 2 `if` check if you could do 1 `if` and return early. What I mean is that this [line](https://github.com/ghiscoding/SlickGrid/blob/bd83368fba49884d905d937180ccde4ce6fff9fb/plugins/slick.rowdetailview.js#L245) can possibly be moved before this [line](https://github.com/ghiscoding/SlickGrid/blob/bd83368fba49884d905d937180ccde4ce6fff9fb/plugins/slick.rowdetailview.js#L238)

Also wondering, where does the `5` come from? I see you use it at a few places in the code but I don't know how you found this magic number `5` (e.g. `rowIndex - 5`, `rowPadding + 5`)? Should we create a global variable to put that `5` if it represent something concrete?

Lastly, just curious, is it still more fluid (faster) compare to your first implementation (`handleOnScrollSaveDetailView`)? Just hope it's still a nice improvement on your side.

**Comment by SatanEnglish** [comment]
First 2 lines you have suggested I had taken the else out cos I was worried but you are **right** as it is actually the same row it cannot be both going out the bottom and exiting the top.

yep swapping those would be right.

The MAGIC 5 yep this is just some number I picked out of the air might be ok to be 1 it is just to determine that the item is at the right end of the viewport. 

![image](https://user-images.githubusercontent.com/2100563/48048202-93f4bc80-e1ff-11e8-95ad-a4efae8246b4.png)

Not sure my picture helps but don't think it needs to be gloabal.

`handleOnScrollSaveDetailView` I'm not sure but at the very least it's easier to figure out whats going on now. I'll have a try and see if it is FASTER when I have a chance.


Filtering wise... Starting to wonder if you can see my comments correctly I've said 3 times now about the fix and that it may not be good enough.

> Filtering kind of works
>
>>PS: Put the last of time before bed to get the filtering updated for you
>
>I put in a case for saving on tab change I think. May need to put on explicitly for on filter call save all >expanded Rows that are in renderRange? (Is this what your thinking here)

I'll put in your suggested fixes. Also wonder if I should move the save part inside somehow.

**Comment by ghiscoding** [comment]
> handleOnScrollSaveDetailView I'm not sure but at the very least it's easier to figure out whats going on now. I'll have a try and see if it is FASTER when I have a chance.

When you test it, make sure to test it without any `console.log` as that will slow down a lot. 

For the magic 5, I just think we should avoid having random number in the code. I said global variable, but I mean global to the plugin and not necessarily available to the user, but at least it's on the top.

> I put in a case for saving on tab change I think. May need to put on explicitly for on filter call save all expanded Rows that are in renderRange? (Is this what your thinking here)

I saw the message before, but I misunderstood it. 

> I'll put in your suggested fixes. Also wonder if I should move the save part inside somehow.

Not sure do understand where the `inside` is? inside of what function?

**Comment by SatanEnglish** [comment]
> For the magic 5, I just think we should avoid having random number in the code.
Yep a global but non user access variable sounds good to be.

> Not sure do understand where the inside is? inside of what function?
My bad not a good explanation by inside I mean within the 'UP' 'Down' if blocks. It may save a couple of extra saves. e.g. if inside 'UP' and 'DOWN' we can figure what direction we are going and only save the appropriate views.
Currently this is saving anything within the buffer area every move of a scroll.

**Comment by SatanEnglish** [comment]
> This makes me think, we might gain even more performance if we keep a flag on each of the _expandedRows and have a bool flag to know it it was already saved or not, then no reason to saving it again, right? Basically add an extra flag on the row, like row._isSaved (a metadata property just like _padding and all the other properties) and if it's already saved then skip the call of saveDetailView().

I'm not sure how you would achieve this.
You don't really know when the view is changed. 
I guess you could have an `onclick` that is on the full expanded row to determine if the user has clicked in that expanded row since last save.

e.g. Cases
As a user I click on a tab in the expanded view (Needs saving)
As a user I am scrolling but I have not changed anything within the expanded row. (No save needed?)
As a user I update an input box inside of expanded row and scroll it out of view. (Needs Saving?)

Basically the condition of if it needs re-saving is my concern here.
I have exposed the save so that someone implementing it can call the save thereself so that maybe enough?

Have a think about what `_isSaved` conditions may look like.

**Comment by ghiscoding** [comment]
Yeah that's true, might be hard to know. You can leave it as it is now then. If you could just update the code for what you said earlier
> Currently this is saving anything within the buffer area every move of a scroll.

Can you also create global variable (in the plugin like I said earlier). Then I think we'll be done.

**Comment by SatanEnglish** [comment]
I'll see if I can find some time tonight to do final polishing.

Any chance you can have a look at the filtering and see if you can find a nicer way to do it?
Maybe an external accessible function `OnFiltering` that saves all expanded items in view port.
That we can call inside filter?

If you don't think it's worth doing that's fine.

**Comment by ghiscoding** [comment]
I added a [commit](https://github.com/ghiscoding/SlickGrid/commit/11fb51f7424cd5f09ecb183dc221334cb6cb4bc6) to add another way of saving the row detail before filtering. I started by adding a new public function `getExpandedRows()` that I can use in the example, and then what I do is simply call the `saveDetailView()` for each of the expanded rows triggered by `keyup` of a filter. This has to happen before the actual filtering, so it's important to put it in a certain order (before the `dataView.refresh();` is fine). I commented out the `saveDetailView` that was inside the `openCity`. 

I think both solution are valid, yours, inside the `openCity` is probably better performance wise since it's only called once after a city change, while my new code will basically save every expanded rows on every `keyup` of every filter. So basically, if user change the city once, but type 3 chars inside the filter, it will call the save 3 times. There's probably some ways to make it more efficient, but I don't think it's important for the demo. User now has 2 options of saving the row detail, so it's still good

**Comment by SatanEnglish** [comment]
I think what you have is over the top for the save by filter.
We only care about the ones in view port. The rest are already in a saved state from scrolling.

**Comment by SatanEnglish** [comment]
Sorry if the last one was bit short it was me typing a quick response between cooking dinner

Can we instead do `_expanded `rows where they are in the `renderedRange `?
This mean when called on filter it will only save the ones open and inside of the renderedRange.
Guess this would change on every type tho.

**Comment by SatanEnglish** [comment]
`function applyTemplateNewLineHeight(item)`
Does some calculation that resize does to it could be refactored. (I'm leaving it no time but letting you know it' there)

Hay may be a couple of days before I get the next chance to change it the last thing I want to do is this.

	  var outterHeight = item[_keyPrefix + 'height'];
	  if(item[_keyPrefix +'sizePadding'] > _options.maxRows){
		  outterHeight = _options.maxRows * rowHeight;
		  item[_keyPrefix +'sizePadding'] = _options.maxRows;
	  }

      // If the padding is now more than the original minRowBuff we need to increase it
      if (_grid.getOptions().minRowBuffer < item[_keyPrefix + 'sizePadding']) {
        // Update the minRowBuffer so that the view doesn't disappear when it's at top of screen + the original default 3
        _grid.getOptions().minRowBuffer = item[_keyPrefix + 'sizePadding'] + 3;
      }

      mainContainer.setAttribute('style', 'height: ' + outterHeight + 'px');
      if (cellItem) cellItem.setAttribute('style', 'max-height: ' + item[_keyPrefix + 'height'] + 'px; top:' + rowHeight + 'px');

This is setting a max rows for expander like this.
![2018-11-08_19-48-30](https://user-images.githubusercontent.com/2100563/48182231-53c44400-e38f-11e8-8770-b35f5d1a7988.gif)


But I'll need to test this at my work end.
This would allow us to set the default to the size of the table to start with so that hopefully it won't do the weird missing rows on scroll.

Anyway I'll have a think and a play at work over the next couple of days.

**Comment by ghiscoding** [comment]
Oh ok, I thought it was only cleaning left to do, was hoping to merge this PR by the weekend. But if you want to add more then that's alright, it will be next week then. Just let me know when it's done

**Comment by SatanEnglish** [comment]
Yer I need to do some tweaking turns out that the max-height and height are on the wrong way round.

E.g.

![image](https://user-images.githubusercontent.com/2100563/48226028-7d1eb780-e403-11e8-8efd-4c24cdf1010e.png)
Blue should have the max-height set so that if our detailView content is bigger than our maxRows it should scroll.
Red should be height so that if it's bigger than the max on blue we get scrolling.

Sorry for the inconvenience would of liked to have this done before my weekend to but better off fixing it now while we still can.

Also I need to fix up naming within the resize method as that isn't helping the confusion.

**Comment by SatanEnglish** [comment]
... Ok now I remember why it's that way round dam it I so cannot win.

The reason that it's was max-height on the outter is so that it can shrink if it's taking up to much rows.

Ok i'll have to have another think I not sure I can have it both ways

**Comment by SatanEnglish** [comment]
FYI I'll push the new version tonight.

Be warned I have overhauled a bunch of the height calculations.
Please give this all a good test over. (Hopefully I've got this all right)

I've also notice a issue that I have no Idea how to resolve maybe you could have a look at it when i've uploaded the new version. (May just be my lack of knowledge on css)

Have changed max-height to min-height to deal with issues around double scroll bars.

I'll message on here again once it's uploaded should be in about 8 hours time

**Comment by SatanEnglish** [comment]
All committed.

Please not the new maxRows cannot be set to less than panelRows else you get a weird issue.
Have a play with pressing resize while in that state. (Won't crash)

Hay do you think it would be worth doing a youtube video on how to use this and add a link into the page?

Also below is examples of what stuff does.
![2018-11-12_17-55-22](https://user-images.githubusercontent.com/2100563/48327541-2aaff600-e6a4-11e8-8d79-59b6ca91b5f5.gif)


![2018-11-12_17-52-14](https://user-images.githubusercontent.com/2100563/48327471-df95e300-e6a3-11e8-8ca6-c1a9654573b3.gif)

There is still an issue were after resize if you close and reopen the tab your resize is lost.
But not 100% sure we can do about that.

**Comment by SatanEnglish** [comment]
Reason for min-height.

When you have tabs assume I clicked on first tab and it's content takes up 60px but the height of panel is 45px. (Therefore you have a scrollbar)
Now assume you click on tab 3 this tab has a height of 100px.
With max-height you end up with 2 Scrollbar due to how tabs work.

With min-height you end up with just 1 Scrollbar as would be expected.

I can do a gif to show you if you would like. The example pre this change had that issue.

**Comment by ghiscoding** [comment]
Ok I just wanted to make sure it doesn't break the default behavior. I didn't have a chance to try it yet, my weekend was quite busy and my desktop computer just stopped working last night (probably & hopefully the power supply)

**Comment by SatanEnglish** [comment]
All good just have a play when you have time next.
If it does make a difference we **MIGHT** be able to make the outter div max-height again.
I stayed with the first thing that worked and didn't try to change that back and test. (finding time is evil)

But as far as I can tell it works the same so no issue there.

Only issue still at the moment is collapsing looses it's height and re-opens at the original panelRows height. (No more time so this will just have to stay this way unless you want to give it half an hour and see if you can deal with it)

**Comment by ghiscoding** [comment]
I reviewed the code and cleaned up a lot, did the rename I talked earlier and removed functions we no longer use. My desktop computer is still broken, so I have to test with my laptop, from what I can so far, it seems to work as before, so it looks fine on my side. 

You could take a last look at it before I submit and merge the PR on the official lib
Thanks

**Comment by ghiscoding** [comment]
@SatanEnglish 
I'm seriously thinking about renaming the `VisibleRange` to `ViewportRange` because the new calculation you've made are more outside of the user's visible range, but I think it's ok to use the `Viewport`. 
So I would rename 
- `onRowOutOfVisibleRange` to `onRowOutOfViewportRange`
- `onRowBackToVisibleRange` to `onRowBackToViewportRange`
    - `outOfVisibleRangeRows` to `rowsOutOfViewport`

**Comment by SatanEnglish** [comment]
All good sorry so bizzy.
That won't affect me to much so all good I'll try review the clean up tomorrow.

**Comment by SatanEnglish** [comment]
Finally had a chance to have a look at your code clean up.

Things that may still won't to clear up.

_viewportRenderedCellCount => _viewportRenderedRowCount
If this will affect you to much don't worry about it just confusing name.

updateOutOfViewportArrayWhenNecessary  => any chance of a little commenting even if it's just one liner?

Apart from that this all looks really good thanks for doing the clean up

**Comment by ghiscoding** [comment]
I've renamed `updateOutOfViewportArrayWhenNecessary` to `syncOutOfViewportArray` which is better and more representative. I removed all reference to `_viewportRenderedCellCount` since it was my code that you stopped using at some point, so there's no reason to keep it. Also made a push in adding comments as much as I could. 

It's now officially a PR [#287](https://github.com/6pac/SlickGrid/pull/287) that will soon be merged, as soon as you say it's all good. Thanks
